### DISCUSSION TITLE
Search and replace possible with regex and varying content?

### DISCUSSION THREAD

**Discussion by sojusnik** [facelessuser/Rummage#463] [Q&A] [answered]
Since I'm not at all into coding, I was wondering if your handy app is able to replace the following text strings with some variations, by using some regex magic and without additional plugins/coding?

My files have quite a few entries with the following pattern:
```
zotero://open-pdf/library/items/ML4WSJ8H?page=3
zotero://open-pdf/library/items/ML4WSJ8H?page=29
zotero://open-pdf/library/items/ML4WSJ8H?page=368
```
mostly in the following form:
`[[zotero://open-pdf/library/items/ML4WSJ8H?page=3|°]]`
which I want to replace into:
`([[zotero://open-pdf/library/items/ML4WSJ8H?page=3|Isak]])`
without touching other entries that look like:
`[[zotero://open-pdf/library/items/ML4WSJ8H?page=3|Summary]]`

The following regex will find the entries with the varying page numbers, but how to replace the string to the format above while preserving the individual page number?
`\[\[zotero://open-pdf/library/items/ML4WSJ8H\?page=\d+\|°\]`

**Comment by facelessuser** [comment]
I'm a bit confused about exactly what you are asking. If you can make it just a little clearer, maybe I can help.

It seems like you want to replace the following, specifically those that have `°`, but avoid cases that don't have `°`?

```
[[zotero://open-pdf/library/items/ML4WSJ8H?page=3|°]]
```
By surrounding it with `()`? And replacing `°` with `Isak`? Or are you saying what you want to replace `°` with something Dynamic, maybe `Isak`, maybe something else dependent upon page number?

**Reply by sojusnik** [reply]
> It seems like you want to replace the following, specifically those that have `°`, but avoid cases that don't have `°`?

Indeed!

> By surrounding it with `()`? And replacing `°` with `Isak`?

Yes, but the tricky part is that text like `[[zotero://open-pdf/library/items/ML4WSJ8H?page=3|Summary]]` should stay untouched.

> Or are you saying what you want to replace `°` with something Dynamic, maybe `Isak`, maybe something else dependent upon page number?

No, the idea is to preserve the different page numbers and replace `°` to `Isak`, while adding the brackets, without altering other text like `[[zotero://open-pdf/library/items/ML4WSJ8H?page=3|Summary]]` or `[[zotero://open-pdf/library/items/ML4WSJ8H?page=624|Chapter]]` and so on.

Just ask ahead if you need further information or other examples.

**Reply by facelessuser** [reply, marked as answer]
1. Use capturing groups in your search pattern`()`. The order in which they are specified can later be represented using numbers. Capture what you want to preserve.
2. Create your replace pattern using the numbers `\1`, `\2`, etc. to replace the part you wish to preserve.

Find:

```regex
(\[\[zotero://open-pdf/library/items/ML4WSJ8H\?page=\d+\|)°(\]\])
```

Replace:

```regex
(\1Isak\2)
```

<img width="679" alt="Screenshot 2024-09-25 at 2 39 09 PM" src="https://github.com/user-attachments/assets/b68038d4-3c5a-4ea3-b9f3-fe680aa3b3bb">

Adapt doing whatever you need.

**Reply by sojusnik** [reply]
This is exactly what I was looking for, thank you very much!
